### PR TITLE
[optimizer] Refactor optimizer

### PR DESCRIPTION
**Pull request by kparichay**
This patch refactors optimizer in the following fashion:
- split optimizer implementations of different types to derived classes of adam and sgd
- create a optimizer_factory to create the optimizer class objs based on its type
  - this can be directly used with ccapi
- OptParam struct has been removed
- applyGradients has been broken down into different methods
- updated associated unittests

In a future PR, I will create subfolders in src/include to manage files better.

See also #199

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/src/optimizer.cpp [hunk 1]
```diff
@@ -34,73 +34,25 @@
...
+  double ll = learning_rate;
 
-int Optimizer::setOptParam(OptParam p) {
-  int status = ML_ERROR_NONE;
-  if (p.learning_rate <= 0) {
-    ml_loge("Error: learning_rate should be grater than 0 (%f)",
-            p.learning_rate);
-    return ML_ERROR_INVALID_PARAMETER;
+  if (decay_steps != -1) {
+    ll = ll * pow(decay_rate, (iteration / decay_steps));
```
Comment (zhoonit) [inline nntrainer/src/optimizer.cpp:46]: `powf` for the precision?

File: nntrainer/src/adam.cpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+int Adam::initialize(std::shared_ptr<Weight> weight_list,
+                     unsigned int num_weights, bool set_tensor) {
+  int status = ML_ERROR_NONE;
+  weight_mv.clear();
+
+  if (set_tensor) {
+    for (unsigned int i = 0; i < num_weights; ++i) {
+      Weight &w = weight_list.get()[i];
+
+      // TODO: only trainable weights must be sent to optimizer
```
Comment (zhoonit) [inline nntrainer/src/adam.cpp:34]: Since trainable weights might not be in the contiguous memory space, this might be the only option to filter weights effectively.

E.g)
weights: `[a, b, c]`, and `b.trainable = false`

File: nntrainer/include/optimizer.h [hunk 3]
```diff
@@ -119,27 +95,19 @@ class Optimizer {
    * @brief     get Learning Rate
    * @retval    Learning rate
    */
-  float getLearningRate() { return popt.learning_rate; };
+  float getLearningRate() { return learning_rate; };
```
Comment (zhoonit) [inline nntrainer/include/optimizer.h:98]: `virtual float getLearningRate()` and change `adam::getLearningRate` to float?

File: nntrainer/src/adam.cpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+      file.seekg(total_size, std::ifstream::cur);
+    }
+  } else {
+    ml_logw("Not loading saved optimizer parameters due to mismatched type");
+  }
+}
+
+void Adam::save(std::ofstream &file) {
+  file.write((char *)&type, sizeof(OptType));
```
Comment (zhoonit) [inline nntrainer/src/adam.cpp:147]:
```suggestion
    Optimizer::save(file)
```

File: nntrainer/src/adam.cpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    status = ML_ERROR_NONE;
+    break;
+  }
+
+  throw_status(status);
+}
+
+void Adam::read(std::ifstream &file) {
+  OptType loaded_type;
+  file.read((char *)&loaded_type, sizeof(OptType));
```
Comment (zhoonit) [inline nntrainer/src/adam.cpp:126]: Optimizer::read(file) will throw so `if(loaded_type == type)` can be removed

File: nntrainer/include/optimizer.h [hunk 6]
```diff
@@ -201,36 +150,82 @@ class Optimizer {
...
+
+  /**
+   * @brief     apply gradient to the given weight
+   * @param[in] weight Weight and gradient set to be updated
+   * @param[in] tensor_idx Idx of this tensor in the tensors list
+   * @param[in] num_weights size of the array
+   * @param[in] iteration nth epoch number
+   * @note weight which is called upon can be assumed to be trainable
+   */
+  virtual void apply_gradient(Weight &weight, int tensor_idx, double updated_lr,
```
Comment (zhoonit) [inline nntrainer/include/optimizer.h:225]: I think double and float are mixed for the learning rate. Could you check once up?

File: nntrainer/src/optimizer.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/src/optimizer.cpp:46]: pow operates on double, powf operates on float.

File: nntrainer/src/adam.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/adam.cpp:34]: I think the weights passed here to be trainable or not is fixed in logic of the layer.
So, layers can possibly two set of containers for weights - trianable and non-trainable
or just index them better.

Reason is that this check is performed in critical path of training - applyGradients()

File: nntrainer/include/optimizer.h (hunk 3 above)
Comment (kparichay) [reply nntrainer/include/optimizer.h:98]: I think you are mixing two things.
There are two functionsin optimizer.h:
`float getLearningRate()` which is a public function
`virtual double getLearningRate(int iteration)` which is a private function and overloaded in adam.h

File: nntrainer/src/adam.cpp (hunk 4 above)
Comment (kparichay) [reply nntrainer/src/adam.cpp:147]: that's my bad, updated.

File: nntrainer/src/adam.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/src/adam.cpp:126]: `if(loaded_type == type)` is still needed. 
Updated `optimizer.read()` to  only throw for unknown types.
it will just load the optimizer type for checking, and then the overload optimizer, if the type matches can load corresponding data.

File: nntrainer/include/optimizer.h (hunk 6 above)
Comment (kparichay) [reply nntrainer/include/optimizer.h:225]: `updated_lr` is intentionally double. Learning rate is handled as a double internally. However, for public view and the initial value are stored as a float but any operation or intermediate results of learning rate are performed as double.

File: nntrainer/src/adam.cpp (hunk 2 above)
Comment (zhoonit) [reply nntrainer/src/adam.cpp:34]: Got it

File: nntrainer/src/optimizer.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/src/optimizer.cpp:46]: No, https://github.com/nnstreamer/nntrainer/pull/633#discussion_r503057084 this explained Thanks.

File: nntrainer/include/optimizer.h (hunk 3 above)
Comment (zhoonit) [reply nntrainer/include/optimizer.h:98]: Got it

File: nntrainer/include/optimizer.h (hunk 6 above)
Comment (zhoonit) [reply nntrainer/include/optimizer.h:225]: Alright

**Review by zhoonit — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM
